File: Wizzem/Wizzem/Controllers/ProfileController/ProfileViewController.m
//
//  ProfileViewController.m
//  Wizzem
//
//  Created by Remi Robert on 26/04/15.
//  Copyright (c) 2015 Remi Robert. All rights reserved.
//

#import <SVProgressHUD.h>
#import <Parse/Parse.h>
#import "ProfileViewController.h"

@interface ProfileViewController () <UITableViewDataSource, UITableViewDelegate>
@property (strong, nonatomic) IBOutlet UITableView *tableView;
@property (nonatomic, strong) UIImageView *imageProfile;
@property (nonatomic, strong) UILabel *numberFollowers;
@property (nonatomic, strong) UILabel *numberFollowing;
@end

@implementation ProfileViewController

- (UIImageView *)imageProfile {
    if (!_imageProfile) {
        _imageProfile = [[UIImageView alloc] init];
        _imageProfile.frame = CGRectMake(0, 0, [UIScreen mainScreen].bounds.size.width / 3, [UIScreen mainScreen].bounds.size.width / 3);
        _imageProfile.layer.cornerRadius = [UIScreen mainScreen].bounds.size.width / 3 / 2;
        
        PFFile *picture = [PFUser currentUser][@"picture"];
        [picture getDataInBackgroundWithBlock:^(NSData *data, NSError *error) {
            if (data && !error) {
                _imageProfile.image = [UIImage imageWithData:data];
            }
        }];
    }
    return _imageProfile;
}

- (void)logout {
    [SVProgressHUD show];
    
    [PFUser logOutInBackgroundWithBlock:^(NSError *PF_NULLABLE_S error) {
        
        [SVProgressHUD dismiss];
        
        if (error) {
            UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Error logout" message:@"Try again" delegate:nil cancelButtonTitle:nil otherButtonTitles:@"Ok", nil];
            [alert show];
        }
        else {
            UIStoryboard *mainStoryboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
            UIViewController *loginController = [mainStoryboard instantiateViewControllerWithIdentifier:@"loginController"];
            if (loginController) {
                [self presentViewController:loginController animated:true completion:nil];
            }
        }
    }];
}

- (void)friendAction {
    [self performSegueWithIdentifier:@"friendController" sender:nil];
}

- (void)followersAction {
    [self performSegueWithIdentifier:@"followingController" sender:nil];
}

- (void)setttingAction {
    UIAlertController *actionMenu = [UIAlertController alertControllerWithTitle:@"Settings" message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    
    [actionMenu addAction:[UIAlertAction actionWithTitle:@"Logout" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
        [self logout];
    }]];
    [actionMenu addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:^(UIAlertAction *action) {
    }]];
    
    
    [self presentViewController:actionMenu animated:true completion:nil];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    
    
    self.numberFollowers.text = [NSString stringWithFormat:@"%ld", [[[PFUser currentUser] objectForKey:@"nbFollower"] integerValue]];
    self.numberFollowing.text = [NSString stringWithFormat:@"%ld", [[[PFUser currentUser] objectForKey:@"nbFollowing"] integerValue]];
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    
    UIView *container = [[UIView alloc] init];
    
    if (section == 0) {
        container.frame = CGRectMake(0, 0, self.view.frame.size.width, 300);
        container.backgroundColor = [UIColor clearColor];
        
        UIView *imageContainer = [[UIView alloc] init];
        imageContainer.frame = CGRectMake(self.view.frame.size.width / 2 - ([UIScreen mainScreen].bounds.size.width / 3 / 2), 20, [UIScreen mainScreen].bounds.size.width / 3, [UIScreen mainScreen].bounds.size.width / 3);
        imageContainer.layer.masksToBounds = true;
        imageContainer.layer.cornerRadius = [UIScreen mainScreen].bounds.size.width / 3 / 2;
        [imageContainer addSubview:self.imageProfile];
        
        UILabel *username = [[UILabel alloc] init];
        username.frame = CGRectMake(0, [UIScreen mainScreen].bounds.size.width / 3 + 40, self.view.frame.size.width, 30);
        username.text = [PFUser currentUser][@"true_username"];
        username.font = [UIFont boldSystemFontOfSize:20];
        username.textAlignment = NSTextAlignmentCenter;
        [username sizeToFit];
        username.frame = CGRectMake(self.view.frame.size.width / 2 - username.frame.size.width / 2, username.frame.origin.y, username.frame.size.width, username.frame.size.height);
        
        UIButton *buttonSettings = [[UIButton alloc] init];
        [buttonSettings setImage:[[UIImage imageNamed:@"settings"] imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate] forState:UIControlStateNormal];
        buttonSettings.tintColor = [UIColor grayColor];
        buttonSettings.frame = CGRectMake(username.frame.origin.x + username.frame.size.width + 10, username.frame.origin.y, username.frame.size.height, username.frame.size.height);
        [buttonSettings addTarget:self action:@selector(setttingAction) forControlEvents:UIControlEventTouchUpInside];

        UIButton *followers = [[UIButton alloc] init];
        followers.frame = CGRectMake([UIScreen mainScreen].bounds.size.width / 3 / 2 - 50, 20 + [UIScreen mainScreen].bounds.size.width / 3 / 2, 100, 100);
        [followers addTarget:self action:@selector(followersAction) forControlEvents:UIControlEventTouchUpInside];
        
        self.numberFollowers = [[UILabel alloc] init];
        self.numberFollowers.frame = CGRectMake(0, 0, 100, 20);
        self.numberFollowers.textAlignment = NSTextAlignmentCenter;
        self.numberFollowers.text = @"0";
        self.numberFollowers.font = [UIFont boldSystemFontOfSize:22];
        
        UILabel *labelTitleFollowers = [[UILabel alloc] init];
        labelTitleFollowers.frame = CGRectMake(0, 20, 100, 20);
        labelTitleFollowers.textAlignment = NSTextAlignmentCenter;
        labelTitleFollowers.text = @"Followers";
        labelTitleFollowers.textColor = [UIColor lightGrayColor];
        [followers addSubview:self.numberFollowers];
        [followers addSubview:labelTitleFollowers];
        
        
        UIButton *following = [[UIButton alloc] init];
        following.frame = CGRectMake(([UIScreen mainScreen].bounds.size.width / 3 * 2) + ([UIScreen mainScreen].bounds.size.width / 3 / 2 - 50), 20 + [UIScreen mainScreen].bounds.size.width / 3 / 2, 100, 30);
        [following addTarget:self action:@selector(friendAction) forControlEvents:UIControlEventTouchUpInside];

        self.numberFollowing = [[UILabel alloc] init];
        self.numberFollowing.frame = CGRectMake(0, 0, 100, 20);
        self.numberFollowing.textAlignment = NSTextAlignmentCenter;
        self.numberFollowing.text = @"0";
        self.numberFollowing.font = [UIFont boldSystemFontOfSize:22];
        
        UILabel *labelTitleFollowing = [[UILabel alloc] init];
        labelTitleFollowing.frame = CGRectMake(0, 20, 100, 20);
        labelTitleFollowing.textAlignment = NSTextAlignmentCenter;
        labelTitleFollowing.text = @"Following";
        labelTitleFollowing.textColor = [UIColor lightGrayColor];
        [following addSubview:self.numberFollowing];
        [following addSubview:labelTitleFollowing];
        
        [container addSubview:imageContainer];
        [container addSubview:followers];
        [container addSubview:following];
        [container addSubview:username];
        [container addSubview:buttonSettings];
        
        self.numberFollowers.text = [NSString stringWithFormat:@"%ld", [[[PFUser currentUser] objectForKey:@"nbFollower"] integerValue]];
        self.numberFollowing.text = [NSString stringWithFormat:@"%ld", [[[PFUser currentUser] objectForKey:@"nbFollowing"] integerValue]];
    }
    else {
        container.backgroundColor = [UIColor whiteColor];
        UISegmentedControl *segment = [[UISegmentedControl alloc] initWithItems:@[@"Passé", @"En cours", @"A venir"]];
        segment.selectedSegmentIndex = 1;
        segment.selected = true;
        segment.frame = CGRectMake(10, 10, self.view.frame.size.width - 20, 30);
        
        container.frame = CGRectMake(0, 0, self.view.frame.size.width, segment.frame.size.height + 20);
        
        [container addSubview:segment];
    }
    return container;
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    if (section == 0) {
        return [UIScreen mainScreen].bounds.size.width / 3 + 20 + 70;
    }
    return 50;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (section == 0) {
        return 0;
    }
    return 200;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"cell"];
    cell.textLabel.text = @"test";
    return cell;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 2;
}

@end

File: Wizzem-iOS/Wizzem-iOS/Controllers/ProfileViewController.swift
//
//  ProfileViewController.swift
//  
//
//  Created by Remi Robert on 18/07/15.
//
//

import UIKit

class ProfileViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {

    @IBOutlet var tableView: UITableView!
    var events = Array<PFObject>()
    var refreshControl: UIRefreshControl!
    var animator: ZFModalTransitionAnimator!
    @IBOutlet var pictureProfile: FLAnimatedImageView!
    @IBOutlet var usernameLabel: UILabel!
    
    @IBAction func editProfile(sender: AnyObject) {
        let alertController = UIAlertController(title: nil, message: nil, preferredStyle: UIAlertControllerStyle.ActionSheet)
        
        let profileAction = UIAlertAction(title: "Voir mon profile", style: UIAlertActionStyle.Default) { (_) -> Void in
            self.performSegueWithIdentifier("detailProfileSegue", sender: nil)
        }
        let logoutAction = UIAlertAction(title: "Deconnexion", style: UIAlertActionStyle.Destructive) { (_) -> Void in
            PFUser.logOutInBackgroundWithBlock({ (error: NSError?) -> Void in
                if error != nil {
                    Alert.error("Erreur lors de la deconnexion")
                }
                else {
                    if let controller = InstanceController.fromStoryboard("loginSignupController") {
                        self.presentViewController(controller, animated: false, completion: nil)
                    }
                }
            })
        }
        let cancelAction = UIAlertAction(title: "Annuler", style: UIAlertActionStyle.Cancel, handler: nil)
        
        alertController.addAction(profileAction)
        alertController.addAction(logoutAction)
        alertController.addAction(cancelAction)
        self.presentViewController(alertController, animated: true, completion: nil)
    }
    
    func editProfile() {
        self.performSegueWithIdentifier("changeProfileSettingSegue", sender: nil)
    }
        
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        tableView.deselectRowAtIndexPath(indexPath, animated: false)
        performSegueWithIdentifier("detailEventSegue", sender: events[indexPath.row])
    }

    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return events.count
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        var cell = tableView.dequeueReusableCellWithIdentifier("momentCell") as! MomentFeedTableViewCell
        
        let currentEvent = events[indexPath.row]
        
        cell.titleMoment.text = nil
        cell.participantLabel.text = nil
        cell.wizzLabel.text = nil
        cell.dataMoment.text = nil
        
        currentEvent.fetchIfNeededInBackgroundWithBlock { (event: PFObject?, _) -> Void in
            if let event = event {
                if let title = event["title"] as? String {
                    cell.titleMoment.text = title
                }
                if let numberParticipant = event["nbParticipant"] as? Int {
                    cell.participantLabel.text = "Avec \(numberParticipant) participants"
                }
                if let wizzNumber = event["nbMedia"] as? Int {
                    cell.wizzLabel.text = "\(wizzNumber)"
                }
                if let city = event["city"] as? String {
                    cell.dataMoment.text = "à \(city)"
                }
            }
        }
        return cell
    }
    
    func fetchData() {
        let querry = PFQuery(className: "Participant")
        querry.whereKey("userId", equalTo: PFUser.currentUser()!)
        
        querry.findObjectsInBackgroundWithBlock { (results: [AnyObject]?, _) -> Void in

            if self.refreshControl.refreshing {
                self.refreshControl.endRefreshing()
            }
            
            if let results = results as? [PFObject] {
                self.events.removeAll(keepCapacity: false)
                for currentParticipant in results {
                    if let event = currentParticipant["eventId"] as? PFObject {
                        self.events.append(event)
                    }
                }
                self.tableView.reloadData()
            }
        }        
    }
    
    func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        return 97
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.usernameLabel.text = nil
        self.pictureProfile.image = nil
        self.pictureProfile.contentMode = UIViewContentMode.ScaleAspectFill
        self.pictureProfile.layer.masksToBounds = true
        self.pictureProfile.layer.cornerRadius = 25
        
        PFUser.currentUser()?.fetchInBackgroundWithBlock({ (user: PFObject?, _) -> Void in
            if let user = user {
                self.usernameLabel.text = user["true_username"] as? String
                if let filePicture = user["picture"] as? PFFile {
                    filePicture.getDataInBackgroundWithBlock({ (data: NSData?, _) -> Void in
                        if let data = data {
                            if let dataType = user["typePicture"] as? String where dataType == "GIF" {
                                dispatch_async(dispatch_get_main_queue(), { () -> Void in
                                    let animatedImage = FLAnimatedImage(animatedGIFData: data)
                                    println("aimated image : \(animatedImage)")
                                    self.pictureProfile.animatedImage = animatedImage
                                })
                            }
                            else {
                                self.pictureProfile.image = UIImage(data: data)
                            }
                        }
                    })
                }
            }
        })
    
        let logo = UIImageView(frame: CGRectMake(8, 44 / 2 - 15, 70, 30))
        logo.image = UIImage(named: "LogoWz")
        logo.backgroundColor = UIColor.clearColor()
        logo.contentMode = UIViewContentMode.ScaleAspectFit
        self.navigationController?.navigationBar.addSubview(logo)

        let titleLabel = UILabel()
        titleLabel.textAlignment = NSTextAlignment.Right
        titleLabel.text = "Mes moments "
        titleLabel.frame.size = CGSizeMake(CGRectGetWidth(UIScreen.mainScreen().bounds) - 40, 40)
        titleLabel.frame.origin = CGPointMake(20, 24)
        titleLabel.font = UIFont(name: "ArialRoundedMTBold", size: 18)!
        self.navigationItem.titleView = titleLabel
        
        let v = UIView(frame: CGRectMake(0, 0, UIScreen.mainScreen().bounds.size.width, 100))
        v.backgroundColor = UIColor.redColor()
        
        refreshControl = UIRefreshControl()
        refreshControl.addTarget(self, action: "fetchData", forControlEvents: UIControlEvents.ValueChanged)
        refreshControl.backgroundColor = UIColor.clearColor()
        refreshControl.tintColor = UIColor(red:0.3, green:0.85, blue:0.39, alpha:1)
        tableView.addSubview(refreshControl)
        
        tableView.registerNib(UINib(nibName: "MomentFeedTableViewCell", bundle: nil), forCellReuseIdentifier: "momentCell")
        
        tableView.dataSource = self
        tableView.delegate = self
        
        fetchData()
        
        navigationController?.navigationBar.setBackgroundImage(UIImage(), forBarPosition: .Any, barMetrics: .Default)
        navigationController?.navigationBar.shadowImage = UIImage()
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "detailEventSegue" {
            (segue.destinationViewController as! DetailMediaViewController).currentEvent = sender as! PFObject
            
            let controller = segue.destinationViewController as! UIViewController
            
            self.animator = ZFModalTransitionAnimator(modalViewController: controller)
            self.animator.dragable = true
            self.animator.bounces = false
            self.animator.direction = ZFModalTransitonDirection.Right
            self.animator.behindViewAlpha = 0.5
            self.animator.behindViewScale = 0.9
            self.animator.transitionDuration = 0.7
            controller.transitioningDelegate = self.animator
        }
        else if segue.identifier == "detailProfileSegue" {
            (segue.destinationViewController as! DetailProfileViewController).user = PFUser.currentUser()!
        }
    }
}
